### PR TITLE
[sdk-traces] SamplingResult constructors accept null parameters

### PR DESCRIPTION
**Pull request by Kielek**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1718#discussion_r1597926848
Design discussion issue #

## Changes

SamplingResult constructor accepts null parameters

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* ~~[ ] Unit tests added/updated~~
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -25,4 +25,8 @@ OpenTelemetry.ReadOnlyFilteredTagCollection.Enumerator.Enumerator() -> void
 OpenTelemetry.ReadOnlyFilteredTagCollection.Enumerator.MoveNext() -> bool
 OpenTelemetry.ReadOnlyFilteredTagCollection.GetEnumerator() -> OpenTelemetry.ReadOnlyFilteredTagCollection.Enumerator
 OpenTelemetry.ReadOnlyFilteredTagCollection.ReadOnlyFilteredTagCollection() -> void
+*REMOVED*OpenTelemetry.Trace.SamplingResult.SamplingResult(OpenTelemetry.Trace.SamplingDecision decision, string! traceStateString) -> void
+*REMOVED*OpenTelemetry.Trace.SamplingResult.SamplingResult(OpenTelemetry.Trace.SamplingDecision decision, System.Collections.Generic.IEnumerable<System.Collections.Generic.KeyValuePair<string!, object!>>! attributes) -> void
```
Comment (reyang) [inline src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt:29]: Are these still needed or can be removed?

Comment (Kielek) [reply src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt:29]: IMO needed. Changes should be propagated to released file while creating release PR.

Comment (CodeBlanch) [reply src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt:29]: If all you are doing is changing nullable annotations, I think it is safe to remove these changes in the "Unshipped" file and just go update the lines in the shipped file. Change "!" to "?" for the things being made nullable. Nullable decorations don't really change the code just analysis of it there's nothing breaking with this change AFAIK.

File: src/OpenTelemetry/CHANGELOG.md [hunk 2]
```diff
@@ -25,6 +25,12 @@
   supported in stable builds.
   ([#5607](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5607))
 
+* `SamplingResult(SamplingDecision decision,
+  IEnumerable<KeyValuePair<string, object>>? attributes)`
+  and `SamplingResult(SamplingDecision decision, string? traceStateString)` constructors
+  accept `null` for `attributes` and `traceStateString` parameters respectively.
+  ([#5614](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5614))
```
Comment (CodeBlanch) [inline src/OpenTelemetry/CHANGELOG.md:32]: I think call it a fix. Suggestion...

* Fixed the nullable annotations for the `SamplingResult` constructors to allow `null` being supplied as `attributes` or `traceStateString` which has always been allowed.

Comment (reyang) [reply src/OpenTelemetry/CHANGELOG.md:32]: +1, or even omit the changelog entry?

File: src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (Kielek) [reply src/OpenTelemetry/.publicApi/Stable/PublicAPI.Unshipped.txt]: Fixed in d1c9071e70a70533cad69ad60c32cb43559d978e

File: src/OpenTelemetry/CHANGELOG.md (hunk 2 above)
Comment (Kielek) [reply src/OpenTelemetry/CHANGELOG.md]: Fixed in d3cf82938a32bf0cb3c484a4390ec65745b48a01

File: src/OpenTelemetry/CHANGELOG.md [hunk 3]
```diff
@@ -25,6 +25,11 @@
   supported in stable builds.
   ([#5607](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5607))
 
+* Fixed the nullable annotations for the SamplingResult constructors
```
Comment (Kielek) [inline src/OpenTelemetry/CHANGELOG.md:28]:
```suggestion
* Fixed the nullable annotations for the `SamplingResult` constructors
```

**Review by reyang — APPROVED** [review]

**Pull request merged**